### PR TITLE
Add new devicePixel ref test

### PR DESCRIPTION
**Pull request by greggman**
What the test does:

It sets the body to have a 2x2 checkerboard pattern fit to devicePixels (1 pixel in the pattern = 1 devicePixel)

It then puts a bunch of children above that in rows and uses `devicePixelContentBoxSize` to compute where each child is. It sets the same pattern for each child and offsets the pattern based on the child's calculated position so the patterns should align. If they don't align because the values returned from `devicePixelContentBoxSize` it should be very clear.

Here's Chrome running the test

スクリーンショット 2022-07-01 午後6 28 34

Here's Firefox running the test

スクリーンショット 2022-07-01 午後6 27 50

Related:
https://github.com/web-platform-tests/wpt/pull/34569

### CODE REVIEW

**Review by atotic — COMMENTED** [review]
I will defer to dholbert on this one. I no longer work on web standards, and have used up my "free review time" for a while.

File: resize-observer/devicepixel2.html
Comment (chrishtr) [inline resize-observer/devicepixel2.html:145]: Can you just achieve the same goal with a minimal number of elements declared in HTML rather than built in script? That will be a better reduced testcase, and also make it more clear what layout you're achieving

**Review by dholbert — CHANGES_REQUESTED** [review]
sorry for the delay here!

Overall, this makes sense to me, as long as we have the guarantee that two adjacent flex items should be drawn with directly-abutting device-pixel edges on their content boxes (which seems like a reasonable assumption).

I've got some readability/clarity notes at this point, but the overall approach seems reasonable to me.

File: resize-observer/devicepixel2-ref.html
Comment (dholbert) [reply resize-observer/devicepixel2-ref.html:29]: to clarify, this is the "angry red icon" I'm talking about -- it only shows up in some github UI, e.g. on the direct view of the commit itself:
image

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:110]: "setPatternInEntries" is a bit of a confusing name, since we're not putting the pattern in the entries.

Maybe rename this to "setPatternsUsingSizeInfo" or something along those lines?  That seems to be what this does, rather than doing something "in the entries".

File: resize-observer/devicepixel2.html
Comment (dholbert) [reply resize-observer/devicepixel2.html:119]: (er, I guess there are two callsites; but the second callsite doesn't actually need this lambda and could be simplified to just `setPattern(backElem,0,0)` given that we don't actually need the width or height.  And as noted elsewhere, `backElem` might want to just disappear entirely for improved consistency with the reference case.)

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:19]: 99.5px feels unnecessarily-magic; consider changing this to 100px (or even 127px [same ballpark, but prime].)

I think your intent is that 99.5px is "weird" and likely to produce bugginess, but 100px actually produces more fully-black areas (and it does so sooner, i.e. in higher-up rows, than 99.5px does on my system at least -- Firefox Nightly with devicePixelRatio = 1)

The fractional size here implicitly suggests that we need a fractional size to trigger the bug in question; but in fact we don't.

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:23]: Let's just condense this to `flex: 1;` -- that's essentially equivalent (and simpler). The trailing "1 auto" is irrelevant and unused in practice, since the matched "inner" elements don't have any contents and so their "auto" size is 0 anyway.

(See https://drafts.csswg.org/css-flexbox-1/#flex-common which has the spec's suggested abbreviated usages for the `flex` shorthand, with "flex: <number>" being the preferred way to divide-the-free-space-equally.)

File: resize-observer/devicepixel2.html
Comment (dholbert) [reply resize-observer/devicepixel2.html:145]: > The flexbox tests this indirectly (because varying number of children varies their locations/size).
[...]
>  I'd prefer to explicitly set location/size.

So, I don't think it's possible to both use this test's same relatively-clean "checkerboard overlaying" strategy (which I like) and to **also** directly/explicitly set the position of any element.  If you directly set an element's position to anything other than the origin, then you can't know for sure whether that element will be painted at an even vs. odd device-pixel-position.  And you have to have that information in order to properly set `background-position` on that element (or its children, etc) to align their checkerboard-image-background with the document's checkerboard-image-background.

(That's kind of the point of this test -- if the cumulative `devicePixelBoxSize` measurements of assumed-to-be-adjacent elements are correct, then we'll be able to correctly predict how to set `background-position` on the various checkerboard elements so that they align properly with the checkerboard behind them and won't stand out).

IMO this setup is pretty reasonable.  I could imagine (but probably wouldn't require) simplifying things by making all the flex containers have only 2 flex-items, both inflexible (`flex: none` with specified `width` and `height`), with a variety of widths on the first items in the flex containers (e.g. ranging from 10.0px, 10.1px, etc, up to 11px), so that the second flex item will end up snapping in different directions between different flex containers.  Then we would set the `background-position` on the second flex item based on its assumed device-pixel x/y coordinates, which we would determine based on the previous flex item's observed size and based on the cumulative sizes of the previous flex containers (as the test currently does).

This would let us use explicit hardcoded sizes, and it would sort-of let us explicitly specify positions as well (via positions being trivially inferrable based on the sizes of prior-siblings).

(We could also achieve a similar setup using floats instead of flexbox, but IMO float layout is hairier than flexbox in simple setups like this, particularly when we're using flexbox pretty trivially as this test does with no complex content-based sizing or alignment.)

File: resize-observer/devicepixel2-ref.html
Comment (greggman) [reply resize-observer/devicepixel2-ref.html:29]: fixed

File: resize-observer/devicepixel2-ref.html
Comment (greggman) [reply resize-observer/devicepixel2-ref.html:20]: fixed

File: resize-observer/devicepixel2-ref.html
Comment (greggman) [reply resize-observer/devicepixel2-ref.html:4]: Fixed. But just FYI, I created this file by copying [`devicepixel.html`](https://github.com/greggman/wpt/blob/master/resize-observer/devicepixel.html)

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:140]: done

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:146]: fixed

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:108]: fixed.

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:119]: fixed

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:19]: fixed

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:23]: done

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:110]: done

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:45]: done

File: resize-observer/devicepixel2-ref.html
Comment (greggman) [reply resize-observer/devicepixel2-ref.html:28]: done

**Comment by greggman** [comment]
Thank you for the review. I think I addressed all the issues you raised. 

I changed the pattern to be 4x4 instead of 2x2. My thinking was 2x2 will fail if off by 1 but succeed if off by 2 or if both x and y are off by 1.  With 4x4 the pattern should fail in more cases (though I expect the normal "bad" case to be off by 1)

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:10]: Is there a reason we're bothering with `front` here at this point?  Can we just get rid of it and append to `body` instead, to simplify a bit more?

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:163]: nit: add newline at end of file

(I think you fixed this for the reference case but not for the testcase)

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:38]: s/4/patternSize/ here maybe, since we've got that named constant?

**Review by dholbert — COMMENTED** [review]
Thanks for addressing the review feedback!  I'll take a further look later and run this against Firefox with the patch (the fix for our bug landed today!)

Dropping some quick nits for now...

**Comment by dholbert** [comment]
Viewing today's updated test-version in Chrome on my regular no-pixel-scaling display, I see "fringes" around various flex items at a handful of zoom levels -- e.g. 110% and 125% at least (the first two Ctrl+ ticks).

Needless to say, those location-specific "fringes" aren't present in the reference case, because the reference case just draws a single pattern across the whole page.  So, the test seems to fail in Chrome under full-page zoom (and presumably at various HiDPI display levels).

Assuming a valid test and implementation, I think we wouldn't expect these fringes to be present... Full-page-zoom does get reflected in `devicePixelRatio`, so the test should theoretically be accounting for it properly in its device pixel arithmetic.  I wonder if this is a test bug (/ if we're assuming too much) or a Chrome bug.

**Comment by greggman** [comment]
wow, that was hard to see. It was a bug in the test. The pattern needed to be offset `-(x % patternSize)`. That bug didn't show when the pattern was 2x2 since positive or negative would be the same but with a 4x4 pattern the offset needs to be negative. Uploaded a fix and I changed the pattern so the issue would have stuck out more.  Also note: Firefox fixed their bug yesterday so Firefox Nightly display this correctly now.

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:10]: so I just tried this and the bug does not repo in firefox if I append to body. I don't know if that still means the test should append to body or not. I honestly don't know the intent of the WPT. In WebGL conformance tests (which I'm very familiar with), we check in tests all the time that provoke bugs in individual browsers.

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html]: done

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html]: done

File: resize-observer/devicepixel2.html
Comment (dholbert) [reply resize-observer/devicepixel2.html:10]: > so I just tried this and the bug does not repo in firefox if I append to body

Thanks for testing that! Definitely want to be sure the bug still repro's in affected versions.

Can you just remove these #front styles? They should be unnecessary, and I confirmed I can still repro the bug in Firefox release with that change.  (Doesn't seem to make a difference.)

(I started to test appending to the body, but that's complicated by the fact that the we iterate over the elements using `children`, and the `<script>` element is one of the body's children right now. :))

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:52]: s/were/where/

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:64]: (optional): I think you could set the background-position (xOff and yOff) to just `-x -y` in device pixels (no modulus operation needed), right?

Essentially, we'd be setting every element's background-position to 0 0 in the "global coordinate system (which for an element at `x,y` is `-x, -y`).  Then it will be trivially aligned/properly-overlaid, as long as we have the correct device-pixel positioning (i.e. as long as we don't have the Firefox bug).

The current modulus setup works too, but feels slightly more complex / less intuitive, IMO.

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:118]: This comment needs a rewrite, if you take my above suggestion.

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:67]: (optional): I think you could set the background-position to `-x -y` in device pixels (i.e. `${xOff * oneDevicePixel}px)` and skip the modulus operations, and get rid of xOff/yOff?

Essentially, we can just set every element's background-position to 0 0 in the "global coordinate system (and for an element at `x,y`, the relative position of the global origin is `-x, -y`).

Then every element's background-pattern will be trivially aligned with the body's background, as long as we have the correct x/y positions here (which is the bug we're testing).

The current modulus setup works too, but feels slightly more complex / less intuitive, IMO.

(I confirmed that I can still repro the bug in Firefox release with `s/xOff/x/` and `s/yOff/y/` here, too, BTW.)

File: resize-observer/devicepixel2-ref.html
Comment (dholbert) [inline resize-observer/devicepixel2-ref.html:21]: You've got a different pixel-pattern here vs. in the testcase, I think by accident? This makes the test fail in all browsers for un-interesting reasons right now. :)  (Though it's hard to see without e.g. using an accessibility tool to scale up the device pixels.)

If I fix this (e.g. using the testcase's setup in the reference case), then I confirmed that the test passes for me in the WPT harness (including with the other changes I suggested -- using `-x -y` in device pixels and removing the `#front` css rule)

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:64]: done

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:67]: done

File: resize-observer/devicepixel2-ref.html
Comment (greggman) [reply resize-observer/devicepixel2-ref.html:21]: done

File: resize-observer/devicepixel2.html
Comment (greggman) [reply resize-observer/devicepixel2.html:118]: done

**Comment by greggman** [comment]
Thanks again for the review.  I think I addressed your latest comments.

**Comment by dholbert** [comment]
Thanks for the fixes!

One more thing to fix: there's a "lint" task that's failing, due to some whitespace-at-the-end-of-various-lines (see the `lint` task in the `Some checks were not successful` section here -- direct link: https://github.com/web-platform-tests/wpt/pull/34682/checks?check_run_id=7383242439

That needs fixing. With that, this might be good, I think? I'll take a final look tomorrow and let you know if I see any final issues. :)

FWIW, I also ran a (slightly tweaked) version of your test (as-of-earlier-today) through Firefox Nightly CI, and I confirmed that it passes on trunk -- e.g. here's the pass result for a MacOS run:
https://treeherder.mozilla.org/logviewer?job_id=384603288&repo=try&lineNumber=19899-19902
...and I'm doing another Try run to confirm that it fails on builds from before the patch landed (which I expect it does based on observing it rendering badly in Firefox release):
https://treeherder.mozilla.org/#/jobs?repo=try&revision=e178b200151fdfe17056da1e7c5cd99da8c53ed3

**Comment by dholbert** [comment]
(There's also a failing task in "Azure Pipelines" but that seems to be some sort of infrastructure issue with Safari Technical Preview failing to be downloaded for a particular platform -- not your problem / not related to this test, I think.

Direct link to the failure there:
https://dev.azure.com/web-platform-tests/b14026b4-9423-4454-858f-bf76cf6d1faa/_apis/build/builds/84592/logs/27
```
2022-07-18T04:53:00.0501510Z + ./wpt install --channel preview --download-only -d . --rename STP safari browser
2022-07-18T04:53:07.8874680Z INFO:install:Now installing safari browser...
[...]
2022-07-18T04:53:08.1433370Z ValueError: no download for 12.4
2022-07-18T04:53:08.1621610Z ##[error]Bash exited with code '1'.
2022-07-18T04:53:08.1715030Z ##[section]Finishing: Install Safari Technology Preview
```

@jgraham do you know who we should notify about that? (Perhaps someone who works on WPT infra stuff and/or Safari/WPT integration?)

**Comment by dholbert** [comment]
(maybe @gsnedders knows what's going on with the Safari Technology Preview azure-pipelines task, and if that's something we need to worry about? [seems like an infrastructure issue, as noted above]

**Comment by gsnedders** [comment]
> (maybe @gsnedders knows what's going on with the Safari Technology Preview azure-pipelines task, and if that's something we need to worry about? [seems like an infrastructure issue, as noted above]

#34875

File: resize-observer/devicepixel2-ref.html
Comment (dholbert) [inline resize-observer/devicepixel2-ref.html:16]: `s/4/patternSize/`

(for consistency with the testcase, where I think you fixed this previously)

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:3]: As atotic noted, "correctness" is sort of undefined for pixel-snapping, so it feels a little like over-promising to say that we're checking for "correct" pixel-snapping.

Maybe better/more-precise to say something like this:
> "Resize Observer's reported device pixel content box size should be consistent with the actual pixel-snapped painted box size"

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:102]: This sentence doesn't quite make sense to me.  Maybe clearer:

> Using the children's device-pixel positions (determined as described above), we then set each child's background-position to the negated version of the position that we computed for that child.  This effectively makes its pattern's origin match the document origin.

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:20]: nit: let's move the closing </body> to the end (as you did for the reference case), to reflect the reality of the actual DOM structure here.

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:114]: maybe move the `elemToDevicePixelSize` declaration inside of the `setPatternsUsingSizeInfo()` function, since that's the only place it's used?

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:83]: This chunk of the comment doesn't match what the test actually does.

In particular, three things to address:
 - The phrase ".outer children" is ambiguous & sounds like it's referring to the .outer elements themselves (it sounds like "children who have class .outer"); and that's confusing because the .outer elements don't get any background at all.  Maybe just rephrase as "inner children"? (since you're talking about the innermost things)

 - The current wording suggests that we set the background-position in a final pass (with the image + size having been set earlier).  But in fact, we set the pattern/size/position all at once.

 - The current wording suggests that we set the body's background first and then handle all of the elements later.  But in fact, we handle the body *last*.

Could you update the wording here to address these?  (In the case of the last item, we could just move the `setPattern(document.body, 0, 0);` to be the first thing that happens in the ResizeObserver callback. It is useful to keep it in the callback, to ensure that that the test will fail if the callback never fires.)

**Review by dholbert — COMMENTED** [review]
Final clarity nits; looks good to me with these changes!

File: resize-observer/devicepixel2.html
Comment (dholbert) [inline resize-observer/devicepixel2.html:123]: Sorry, one more thing...  It occurs to me that "outer" and "front" are sort of confusing names at this point.  In particular, the seem ~backwards:
 - "outer" is not the outermost element, but rather "front" is the outermost thing.
 - "front" is the container for all of the page's content, which puts it in **back** of all the page's content (not really "in front")

Maybe clearer if we renamed "outer" to "line" or "row", and then rename "front" to "wrapper"?

**Review by dholbert — COMMENTED** [review]
Looks good, aside from two small typos - thanks for all the fixups & getting this over the finish line! I think we can merge once these last nits are addressed.

**Review by dholbert — APPROVED** [review]

**Comment by dholbert** [comment]
I clicked the button to merge; hopefully everything works properly. :)

Thanks again for the Firefox bug report and for the thorough test!

**Comment by dholbert** [comment]
wpt.fyi now has this test deployed, and shows it passing on all browsers; hooray!
https://wpt.fyi/results/resize-observer/devicepixel2.html

**Pull request merged**
